src/hotspot/cpu/ppc/gc/z/zBarrierSetAssembler_ppc.cpp

```diff
@@ -467,19 +471,34 @@ class ZSaveLiveRegisters {
           reg_save_index++;
 
           if (action == ACTION_SAVE) {
-            _masm->stfd(fp_reg, (intptr_t) -reg_save_index * BytesPerWord, R1_SP);
+            _masm->stfd(fp_reg, offset - reg_save_index * BytesPerWord, R1_SP);
           } else if (action == ACTION_RESTORE) {
-            _masm->lfd(fp_reg, (intptr_t) -reg_save_index * BytesPerWord, R1_SP);
+            _masm->lfd(fp_reg, offset - reg_save_index * BytesPerWord, R1_SP);
           } else {
-            fatal("Sanity");
+            assert(action == ACTION_COUNT_ONLY, "Sanity");
           }
         }
       } else if (vm_reg->is_ConditionRegister()) {
         // NOP. Conditions registers are covered by save_LR_CR
+      } else if (vm_reg->is_VectorSRegister()) {
+        assert(SuperwordUseVSX, "or should not reach here");
+        VectorSRegister vs_reg = vm_reg->as_VectorSRegister();
+        if (vs_reg->encoding() >= VSR32->encoding() && vs_reg->encoding() <= VSR51->encoding()) {
```

Comment:
Why VSR32 as lower bound? I read in ppc.ad
```
1st 32 VSRs are aliases for the FPRs wich are already defined above.
```
Could you please help and explain what this means?

Why VSR51 as upper bound?

I'd suggest to update the comment in register_ppc.hpp and explain the vector scalar registers.
What is the difference between vector and vector scalar registers?

Reply:
Thanks for looking at it!
VSRs are not separate registers. They contain the regular FPRs (mapped to 0-31) and VRs (mapped to 32-63). FPRs are managed separately while the VRs are not defined elsewhere in the ppc.ad file. There are instructions which operate on VSRs and can access FPRs and VRs. This was tricky to implement in hotspot ([JDK-8188139](https://bugs.openjdk.org/browse/JDK-8188139) and many follow-up fixes).
Only the VRs VR0-VR19 are volatile (see register_ppc.hpp), so only these ones need spilling. (Same is done for other register types.)
VR0-VR19 = VSR32-VSR51
Note that only these ones are currently used by C2 (see `reg_class vs_reg` in ppc.ad). Reason is that we currently don't preserve the non-volatile ones in the Java entry frame.

Reply:
> Thanks for looking at it! VSRs are not separate registers. They contain the
> regular FPRs (mapped to 0-31) and VRs (mapped to 32-63). FPRs are managed
> separately while the VRs are not defined elsewhere in the ppc.ad file.

Thanks. I think this should be better explained in register_ppc.hpp.

> There are instructions which operate on VSRs and can access FPRs and VRs. This
> was tricky to implement in hotspot
> ([JDK-8188139](https://bugs.openjdk.org/browse/JDK-8188139) and many follow-up
> fixes). Only the VRs VR0-VR19 are volatile (see register_ppc.hpp), so only
> these ones need spilling. (Same is done for other register types.) VR0-VR19 =
> VSR32-VSR51
> Note that only these ones are currently used by C2 (see `reg_class
> vs_reg` in ppc.ad). Reason is that we currently don't preserve the
> non-volatile ones in the Java entry frame.

I see. VSR52-VSR64 are declared SOC in ppc.ad. Shouldn't they be SOE then?

Reply:
I've added a comment to register_ppc.hpp.
Right, they should be SOE. Changed. Note that this doesn't have any effect because the SOE registers are not allocated by C2. But should get fixed to avoid confusion and for possible future usage.